Add shutdown to EventListener

dominikzalewski opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

@PreDestroy and @PostCreate annotations are processed on EventListener methods.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( X ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

findepi commented on core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:
```diff
@@ -83,6 +86,7 @@ public void addEventListener(EventListener eventListener)
         requireNonNull(eventListener, "EventListener is null");
 
         providedEventListeners.add(eventListener);
+        lifeCycleManager.addInstance(eventListener);
```
i think it's not unreasonable for `io.trino.spi.eventlistener.EventListener` to be informed about system shutdown

if we want that, this should be an explicit method on that interface though, not some implicit annotation-reading which is hard to discover by interface implementors, and may not even work across classloaders.

kokosing replied:
Let's follow `io.trino.spi.connector.Connector#shutdown` practice

wendigo replied:
Yes. It'd be better to dispatch the shutdown event from the `EventListenerManager`. It should have @PreDestroy method that will call all event listeners shutdown method

electrum replied:
This wouldn't work anyway, as the annotations in the plugin are in a different class loader than the engine.

wendigo commented on core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:
```diff
@@ -62,11 +63,13 @@ public class EventListenerManager
     private final TimeStat queryCreatedTime = new TimeStat(MILLISECONDS);
     private final TimeStat queryCompletedTime = new TimeStat(MILLISECONDS);
     private final TimeStat splitCompletedTime = new TimeStat(MILLISECONDS);
+    protected final LifeCycleManager lifeCycleManager;
```
why is it protected?

dominikzalewski replied:
I needed that in TestingLifecycleManager that inherits. But this is going to change now anyway.

electrum reviewed (changes requested):
We should add an explicit `shutdown` method, like `Connector`

dominikzalewski commented:
Changed as requested:
- EventListener.shutdown() also calls `close` if it is `AutoCloseable` -> let me know whether that's OK with you. If not, I'll remove
- Added unit tests to verify that the call is forwarded from manager to listener

wendigo commented on core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java:
```diff
@@ -34,4 +34,12 @@ default boolean requiresAnonymizedPlan()
     }
+
+    default void shutdown()
+            throws Exception
```
I'd just make it:
```
default void shutdown() {}
```

wendigo commented on core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java:
```diff
@@ -34,4 +34,12 @@ default boolean requiresAnonymizedPlan()
+    default void shutdown()
+            throws Exception
+    {
+        if (this instanceof AutoCloseable autoCloseable) {
```
AutoCloseable is not needed. Event Listener isn't used in try-with-resources block anyway

wendigo commented:
And this failure is related:

```
java.lang.AssertionError: io.trino.plugin.base.classloader.ClassLoaderSafeEventListener does not override [public default void io.trino.spi.eventlistener.EventListener.shutdown() throws java.lang.Exception]
```

wendigo commented on core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java:
```diff
@@ -218,4 +219,17 @@ public TimeStat getSplitCompletedTime()
+    @PreDestroy
+    public void shutdown()
+    {
+        for (EventListener listener : configuredEventListeners.get()) {
```
If we are shutting down event listener manager than I think we should actually make it deconfigure existing listeners.

This line should be:

```
for (EventListener listener : configuredEventListeners.getAndSet(List.of())) {
```

which returns previously configured event listeners and swaps them with empty list.

@electrum @findepi @kokosing wdyt?

wendigo commented on core/trino-spi/src/main/java/io/trino/spi/eventlistener/EventListener.java:
```diff
@@ -34,4 +34,8 @@ default boolean requiresAnonymizedPlan()
         return false;
     }
+
+    default void shutdown()
```
nit:

```
/**
 * Shutdown the event listener by releasing any held resources such as
 * threads, sockets, etc. After this method is called,
 * no methods will be called on the event listener.
 */
default void shutdown() {}
```

wendigo commented:
LGTM @dominikzalewski. Thanks!
